packages/medusa/src/services/cart.ts

```diff
@@ -674,6 +679,154 @@ class CartService extends TransactionBaseService {
+                  .join(
+                    " - "
+                  )}] must belongs to the sales channel on which the cart has been created.`
+              )
+            }
+          }
+        }
+
+        const lineItemServiceTx =
+          this.lineItemService_.withTransaction(transactionManager)
+        const inventoryServiceTx =
+          this.inventoryService_.withTransaction(transactionManager)
+
+        const existingItems = await lineItemServiceTx.list(
+          {
+            cart_id: cart.id,
+            variant_id: In([items.map((item) => item.variant_id)]),
+            should_merge: true,
+          },
+          { select: ["id", "metadata", "quantity"] }
+        )
+
+        const existingItemsVariantMap = new Map()
+        existingItems.forEach((item) => {
+          existingItemsVariantMap.set(item.variant_id, item)
+        })
+
+        const lineItemsToCreate: LineItem[] = []
+        const lineItemsToUpdate: { [id: string]: LineItem }[] = []
+        for (const item of items) {
+          let currentItem: LineItem | undefined
+
+          const existingItem = existingItemsVariantMap.get(item.variant_id)
+          if (item.should_merge) {
+            if (existingItem && isEqual(existingItem.metadata, item.metadata)) {
+              currentItem = existingItem
+            }
+          }
+
+          // If content matches one of the line items currently in the cart we can
+          // simply update the quantity of the existing line item
+          item.quantity = currentItem
+            ? (currentItem.quantity += item.quantity)
+            : item.quantity
+
+          await inventoryServiceTx.confirmInventory(
+            item.variant_id,
+            item.quantity
+          )
+
+          if (currentItem) {
+            lineItemsToUpdate[currentItem.id] = {
+              quantity: item.quantity,
+              has_shipping: false,
+            }
+          } else {
+            // Since the variant is eager loaded, we are removing it before the line item is being created.
+            delete (item as Partial<LineItem>).variant
+            item.has_shipping = false
+            item.cart_id = cart.id
+            lineItemsToCreate.push(item)
+          }
+        }
+
+        // Update all items that needs to be updated
+        if (Object.keys(lineItemsToUpdate).length) {
+          await Promise.all(
+            Object.keys(lineItemsToUpdate).map(async (variantId) => {
+              return await lineItemServiceTx.update(
+                variantId,
+                lineItemsToUpdate[variantId]
+              )
+            })
+          )
+        }
+
+        // Create all items that needs to be created
+        await lineItemServiceTx.create(lineItemsToCreate)
+
+        await lineItemServiceTx
+          .update(
+            {
+              cart_id: cartId,
+              has_shipping: true,
+            },
+            { has_shipping: false }
+          )
+          .catch(() => void 0)
```

Comment:
question: Why do we fail silently here? If an error happens, the line items will be in an invalid state unknown to the consumer.

Reply:
What I could do is checking for that specific error and throw otherwise. wdyt @olivermrbl

Reply:
I would rather be safe too and check for that specific error 💪

Reply:
Also, one question i am asking myself, why do the update method throw when there is no element found? I would have thought that we can try to update something and if there is nothing to update then we do nothing. Like and sql query which does not perform an update if there is no tuple. Wdyt?

Reply:
I see your point. Let's keep that in mind for a separate PR to avoid having it block this one 👍